File: neet_code_150/src/stack/gen_par.rs
struct Solution {}

// 解けなかった😭
// n: 1 ~ 8
impl Solution {
    pub fn gen_par(n: i32) -> Vec<String> {
        let mut result: Vec<String> = vec![];

        for i in 0..n {}

        result
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        // Using the function stack instead of an explicitly allocated Vec
        let mut res: Vec<String> = vec![];

        // open, closeを両方ともnからスタートして0に等しくなるまで減らしながら再帰
        fn backtrack(res: &mut Vec<String>, s: String, open: i32, close: i32) {
            if open == 0 && close == 0 {
                res.push(s);
                return;
            }
            if open == close {
                backtrack(res, s.clone() + "(", open - 1, close);
            } else {
                if open > 0 {
                    backtrack(res, s.clone() + "(", open - 1, close);
                }
                if close > 0 {
                    backtrack(res, s.clone() + ")", open, close - 1);
                }
            }
        }

        backtrack(&mut res, String::from(""), n, n);
        res
    }
}

// C++の模範解答
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut res: Vec<String> = vec![];
        Self::generate(n, 0, 0, "".to_string(), &mut res);

        res
    }

    // open, closeを両方とも0からスタートしてnに等しくなるまで増やしながら再帰
    fn generate(n: i32, open: i32, close: i32, str: String, result: &mut Vec<String>) {
        if open == n && close == n {
            result.push(str);
            return;
        }

        if open < n {
            Self::generate(n, open + 1, close, str.clone() + "(", result);
        }
        if open > close {
            Self::generate(n, open, close + 1, str.clone() + ")", result);
        }
    }
}

fn main() {
    let case_1 = 3; // "[((()))","(()())","(())()","()(())","()()()"]
    let case_2 = 1; // "["()"]
                    //
    println!("case_1: {:?}", SolutionAns::generate_parenthesis(case_1));
    println!("case_2: {:?}", SolutionAns::generate_parenthesis(case_2));

    println!("case_1: {:?}", SolutionAnsCpp::generate_parenthesis(case_1));
    println!("case_2: {:?}", SolutionAnsCpp::generate_parenthesis(case_2));
}

File: neet_code_150_cpp/stack/gen_par.cpp
#include <vector>
#include <string>
#include <iostream>

// Rustの模範解答より
class SolutionAnsRust{
    public:
        std::vector<std::string> gen_par(int n) {
            std::vector<std::string> res;
            backtrack(res, "", n, n);
            return res;
        }
    private:
        // open, closeを両方ともnからスタートして0に等しくなるまで減らしながら再帰
        void backtrack(
                std::vector<std::string>& res, 
                std::string s,
                int open,
                int close) {
            if (open == 0 && close == 0) {
                res.push_back(s);
                return;
            }

            if (open == close) {
                backtrack(res, s + "(", open - 1, close);
            } else {
                if (open > 0) {
                    backtrack(res, s + "(", open - 1, close);
                }
                if (close > 0) {
                    backtrack(res, s + ")", open, close - 1);
                }
            }
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<std::string> generateParenthesis(int n) {
            std::vector<std::string> result;
            generate(n, 0, 0, "", result);
            return result;
        }
    private:
        // open, closeを両方とも0からスタートしてnに等しくなるまで増やしながら再帰
        void generate(int n, int open, int close, std::string str, std::vector<std::string>& result) {
            if (open == n && close == n) {
                result.push_back(str);
                return;
            }
            if (open < n) {
                generate(n, open + 1, close, str + '(', result);
            }
            if (open > close) {
                generate(n, open, close + 1, str + ')', result);
            }
        }
};

int main() {
    SolutionAnsRust s_1;

    auto res_1 = s_1.gen_par(3);
    for(const auto& s: res_1) {
        std::cout << s << " ";
    }
    std::cout << std::endl;

    auto res_2 = s_1.gen_par(2);
    for(const auto& s: res_2) {
        std::cout << s << " ";
    }
    std::cout << std::endl;

    auto res_3 = s_1.gen_par(1);
    for(const auto& s: res_3) {
        std::cout << s << " ";
    }
    std::cout << std::endl;


    SolutionAns s_ans;

    auto res_ans_1 = s_ans.generateParenthesis(3);
    for(const auto& s: res_ans_1) {
        std::cout << s << " ";
    }
    std::cout << std::endl;

    auto res_ans_2 = s_ans.generateParenthesis(2);
    for(const auto& s: res_ans_2) {
        std::cout << s << " ";
    }
    std::cout << std::endl;

    auto res_ans_3 = s_ans.generateParenthesis(1);
    for(const auto& s: res_3) {
        std::cout << s << " ";
    }
    std::cout << std::endl;
}
